Samples/AdapterSelection/AdapterSelection/cpp/main.cpp

```diff
@@ -51,42 +47,24 @@ int main(int argc, char* argv[])
 		spAdapter = nullptr;
 	}
 
+    LearningModelDevice device = nullptr;
 	if (i == 0) {
-		printf("There are no available adapters\n");
-		return -1;
-	}
-
-	// user selects adapter
-	printf("Please enter the index of the adapter you want to use...\n");
-	int selectedIndex;
-
-	while (!(cin >> selectedIndex) || selectedIndex < 0 || selectedIndex >= i) {
-		cin.clear();
-		cin.ignore(numeric_limits<streamsize>::max(), '\n');
-		printf("Invalid index, please try again.\n");
-	}
-
-	printf("Selected adapter at index %d\n", selectedIndex);
-
-	// create D3D12Device
-	com_ptr<IUnknown> spIUnknownAdapter;
-	spFactory->EnumAdapters(selectedIndex, spAdapter.put());
-	spAdapter->QueryInterface(IID_IUnknown, spIUnknownAdapter.put_void());
-	com_ptr<ID3D12Device> spD3D12Device;
-	D3D12CreateDevice(spIUnknownAdapter.get(), D3D_FEATURE_LEVEL_12_0, _uuidof(ID3D12Device), spD3D12Device.put_void());
-
-	// create D3D12 command queue from device
-	D3D12_COMMAND_QUEUE_DESC queueDesc = {};
-	queueDesc.Flags = D3D12_COMMAND_QUEUE_FLAG_NONE;
-	queueDesc.Type = D3D12_COMMAND_LIST_TYPE_DIRECT;
-	com_ptr<ID3D12CommandQueue> spCommandQueue;
-	spD3D12Device->CreateCommandQueue(&queueDesc, IID_PPV_ARGS(spCommandQueue.put()));
-
-	// create LearningModelDevice from command queue	
-	com_ptr<ILearningModelDeviceFactoryNative> dFactory =
-		get_activation_factory<LearningModelDevice, ILearningModelDeviceFactoryNative>();
-	com_ptr<::IUnknown> spLearningDevice;
-	dFactory->CreateFromD3D12CommandQueue(spCommandQueue.get(), spLearningDevice.put());
+		printf("There are no available adapters, running on CPU...\n");

```

Comment:
how did you test this? do you have a device with no GPU?

Reply:
you want to do some more testing on this, since DXGI will most likely return to you the basic render device (warp) .   which we don't use for ML.

[https://docs.microsoft.com/en-us/windows/desktop/api/d3d12/nf-d3d12-d3d12createdevice](https://docs.microsoft.com/en-us/windows/desktop/api/d3d12/nf-d3d12-d3d12createdevice)


---
In reply to: [240405112](https://github.com/Microsoft/Windows-Machine-Learning/pull/107#discussion_r240405112) [](ancestors = 240405112)

Reply:
oh yes, it does return that, and that is the one that runs really slow like I mentioned. Should I not include that in the list of adapters the user can choose?

Reply:
right, you want to filter out WARP.  don't offer that as a choice.

---
In reply to: [240839063](https://github.com/Microsoft/Windows-Machine-Learning/pull/107#discussion_r240839063) [](ancestors = 240839063)

Reply:
https://docs.microsoft.com/en-us/windows/desktop/direct3ddxgi/d3d10-graphics-programming-guide-dxgi
I filter out software drivers using both the adapter description DXGI_ADAPTER_FLAG_SOFTWARE flag and the vendorid and deviceid. I think this should be sufficient to guarantee that all enumerated adapters are valid GPUs.

Reply:
Confirmed that when run on a vm without access to GPUs the software adapter is correctly suppressed and execution will auto fallback to CPU.